pilot/pkg/proxy/envoy/v2/ads.go

```diff
@@ -609,7 +635,24 @@ func (s *DiscoveryServer) AdsPushAll(version string) {
 
 	pendingPush := int32(len(tmpMap))
 
+	// Experimental throttles for the push, to deal with memory spikes.
+	// If it works will be replaced with a rate control.
+	pushThrottleCountEnv := os.Getenv("PILOT_PUSH_THROTTLE_COUNT")
+	pushThrottle := 0
+	if pushThrottleCountEnv != "" {
+		var err error
+		pushThrottle, err = strconv.Atoi(pushThrottleCountEnv)
+		if err != nil {
+			adsLog.Warnf("Invalid push throttle %s", pushThrottleCountEnv)
+		}
+	}
+
+	i := 0
 	for _, c := range tmpMap {
+		i++
+		if pushThrottle > 0 && i%pushThrottle == 0 {
+			time.Sleep(100 * time.Millisecond)
+		}
```

Comment:
How is this going to help pilot deal with memory spikes , by delaying pushes to Envoy ?

Reply:
Yes, right now on change we start pushing to all sidecars at once - and all start doing yaml-to-proto and heavy mem allocation. By sleeping after each 50 sidecars (for example) we reduce memory pressure a bit. 

Eventually, as discussed, we'll do smarter sequencing here, like wait for a set of canaries and add rate control.